// @typegoose/typegoose/lib/internal/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoValidClass = exports.NotAllVPOPElementsError = exports.NotStringTypeError = exports.NotNumberTypeError = exports.InvalidTypeError = exports.InvalidPropError = void 0;
const utils_1 = require("./utils");
class InvalidPropError extends Error {
    constructor(targetName, key) {
        super(`Property "${targetName}.${key}" has no buildSchema!`);
    }
}
exports.InvalidPropError = InvalidPropError;
class InvalidTypeError extends Error {
    constructor(targetName, key, Type) {
        super(`"${targetName}.${key}"'s Type is invalid! Type is: "${Type}"`);
    }
}
exports.InvalidTypeError = InvalidTypeError;
class NotNumberTypeError extends Error {
    constructor(targetName, key, enumKey, enumValue) {
        if (enumKey || enumValue) {
            super(`Typeof "${targetName}.${key}" is "Number", value is undefined/null or does not have a reverse mapping!\n`
                + `Encountered with property: ${enumKey}.${typeof enumValue}`);
        }
        else {
            super(`Type of "${targetName}.${key}" property is not a number.`);
        }
    }
}
exports.NotNumberTypeError = NotNumberTypeError;
class NotStringTypeError extends Error {
    constructor(targetName, key, enumKey, enumValue) {
        if (enumKey || enumValue) {
            super(`Typeof "${targetName}.${key}" is "String", enum is not only Strings!\n`
                + `Encountered with property: ${enumKey}.${typeof enumValue}`);
        }
        else {
            super(`Type of "${targetName}.${key}" property is not a string.`);
        }
    }
}
exports.NotStringTypeError = NotStringTypeError;
/** Not All Virtual Populate Elements Error */
class NotAllVPOPElementsError extends Error {
    constructor(name, key) {
        super(`"${name}.${key}" has not all needed Virtual Populate Options! Needed are: ${utils_1.allVirtualoptions.join(', ')}`);
    }
}
exports.NotAllVPOPElementsError = NotAllVPOPElementsError;
class NoValidClass extends TypeError {
    constructor(cl) {
        super(`"${cl}" is not a function(/constructor)!`);
    }
}
exports.NoValidClass = NoValidClass;
